lib/OpenQA/Schema/Result/Jobs.pm

```diff
@@ -1955,6 +1955,29 @@ sub descendants ($self, $limit = -1) {
     $self->{_descendants} = $sth->fetchrow_array;
 }
 
+sub incomplete_ancestors ($self, $limit = -1) {
+    my $ancestors = $self->{_incomplete_ancestors};
+    return $ancestors if defined $ancestors;
+    my $sth = $self->result_source->schema->storage->dbh->prepare(<<~'EOM');
+    with recursive orig_id as (
+      select ? as orig_id, 0 as level, ? as orig_result, ? as orig_reason
+        union all
+      select id as orig_id, orig_id.level + 1 as level, result as orig_result, reason as orig_reason
+        from jobs join orig_id on orig_id.orig_id = jobs.clone_id
+        where result = 'incomplete' and (? < 0 or level < ?))
+      select orig_id, level, orig_reason from orig_id order by level asc;
+    EOM
+    $sth->bind_param(1, $self->id, SQL_BIGINT);
+    $sth->bind_param(2, $self->result, SQL_VARCHAR);
+    $sth->bind_param(3, $self->reason, SQL_VARCHAR);
+    $sth->bind_param(4, $limit, SQL_BIGINT);
+    $sth->bind_param(5, $limit, SQL_BIGINT);
+    $sth->execute;
+    $ancestors = $sth->fetchall_arrayref;
+    shift @$ancestors;
+    $self->{_incomplete_ancestors} = $ancestors;
+}
+
```

Comment:
I must say this becomes rather complicated and the file lib/OpenQA/Schema/Result/Jobs.pm is already unbearably large so I hope somebody has good hints for refactoring. Not necessarily this method but at least somewhere in this file.

Reply:
At least I extracted the `_reason_and_result` method out of `done`, although I'm quite unhappy with the name.

The module has some methods that don't necessarily have to live in Result::Jobs, e.g. `create_asset` which only needs the job id. But the above function is one that clearly belong there IMHO.

No idea how to refactor yet.

Reply:
Not sure how to split a package (acting as a "class") into multiple files in Perl. If that was possible we could e.g. move all code related to cloning/duplicating into a separate file. (Although I must say that I find the file size manageable and I don't think that splitting makes things much easier to find.)